django/conf/service_urls/base.py

```diff
@@ -0,0 +1,89 @@
+import re
+from urllib import parse
+
+
+class Service:
+    validation = re.compile(r'^(?P<scheme>\S+)://\S*')
+
+    def config_from_url(self, engine, scheme, url):
+        raise NotImplementedError('')
+
+    def __init__(self):
+        self._schemes = {}
+
+    def validate(self, data):
+        match = self.validation.match(data)
+        return match.groups()[0] if match else None
+
+    def _parse(self, data):
+        if isinstance(data, str):
+            scheme = self.validate(data)
+            if scheme is None:
+                raise ValueError('{dsn} is invalid, only full dsn urls (scheme://host...) allowed'.format(dsn=data))
+            try:
+                _scheme = self._schemes[scheme]
+            except KeyError:
+                raise ValueError('{scheme}:// scheme not registered'.format(scheme=scheme))
+            callback, engine = _scheme['callback'], _scheme['engine']
+            return callback(self, engine, scheme, data)
+        return data
+
+    def parse(self, data):
+        if isinstance(data, dict):
+            return {k: self._parse(v) for k, v in data.items()}
+        return self._parse(data)
+
+    @staticmethod
+    def parse_url(url, *, multiple_netloc=False):
+        """
+        A method to parse URLs into components that handles quirks
+        with the stdlib urlparse, such as lower-cased hostnames.
+        Also parses querystrings into typed components.
+        """
+        # This method may be called with an already parsed URL
+        if isinstance(url, dict):
+            return url
```

Comment:
Can you explain why this method may be called with an already parsed URL ? And is this something you really want to support ? After all, the caller could also do `url = url if isinstance(url, dict) else Service.parse_url(url)`. Note that i'm not defending one pattern or the other, but just wondering about how you made your choice.